client/download.go

```diff
@@ -17,81 +17,124 @@ import (
+// MMDB file you have previously downloaded. If you don't yet have one
+// downloaded, this can be "". This is used to know if an update is available
+// and avoid consuming resources if there is not.
+//
+// If the current MD5 checksum matches what the server currently has, no
+// download is performed.
+func (c Client) Download(
 	ctx context.Context,
-	editionID string,
-	hash string,
-) (result *ReadResult, err error) {
-	edition, err := r.getMetadata(ctx, editionID)
+	editionID,
+	md5 string,
+) (DownloadResponse, error) {
+	metadata, err := c.getMetadata(ctx, editionID)
 	if err != nil {
-		return nil, err
+		return DownloadResponse{}, err
 	}
 
-	if edition.MD5 == hash {
-		if r.verbose {
-			log.Printf("No new updates available for %s", editionID)
-		}
-		return &ReadResult{EditionID: editionID, OldHash: hash, NewHash: hash}, nil
+	if metadata.MD5 == md5 {
+		return DownloadResponse{
+			Reader:          io.NopCloser(strings.NewReader("")),
+			UpdateAvailable: false,
+		}, nil
+	}
+
+	reader, modifiedTime, err := c.download(ctx, editionID, metadata.Date)
+	if err != nil {
+		return DownloadResponse{}, err
 	}
 
-	date := strings.ReplaceAll(edition.Date, "-", "")
+	return DownloadResponse{
+		LastModified:    modifiedTime,
+		MD5:             metadata.MD5,
+		Reader:          reader,
+		UpdateAvailable: true,
+	}, nil
+}
+
+const downloadEndpoint = "%s/geoip/databases/%s/download?"
+
+func (c *Client) download(
+	ctx context.Context,
+	editionID,
+	date string,
+) (io.ReadCloser, time.Time, error) {
+	date = strings.ReplaceAll(date, "-", "")
 
 	params := url.Values{}
 	params.Add("date", date)
 	params.Add("suffix", "tar.gz")
 
-	escapedEdition := url.PathEscape(edition.EditionID)
-	requestURL := fmt.Sprintf(downloadEndpoint, r.path, escapedEdition) + params.Encode()
+	escapedEdition := url.PathEscape(editionID)
+	requestURL := fmt.Sprintf(downloadEndpoint, c.endpoint, escapedEdition) + params.Encode()
 
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL, nil)
 	if err != nil {
-		return nil, fmt.Errorf("creating download request: %w", err)
+		return nil, time.Time{}, fmt.Errorf("creating download request: %w", err)
 	}
 	req.Header.Add("User-Agent", "geoipupdate/"+vars.Version)
-	req.SetBasicAuth(strconv.Itoa(r.accountID), r.licenseKey)
+	req.SetBasicAuth(strconv.Itoa(c.accountID), c.licenseKey)
 
-	response, err := r.client.Do(req)
+	response, err := c.httpClient.Do(req)
 	if err != nil {
-		return nil, fmt.Errorf("performing download request: %w", err)
+		return nil, time.Time{}, fmt.Errorf("performing download request: %w", err)
 	}
 	// It is safe to close the response body reader as it wouldn't be
 	// consumed in case this function returns an error.
 	defer func() {
 		if err != nil {
+			// TODO(horgh): Should we fully consume the body?
 			response.Body.Close()
 		}
 	}()
 
 	if response.StatusCode != http.StatusOK {
+		// TODO(horgh): Should we fully consume the body?
```

Comment:
Yeah I'm not sure what drawback  there would be to using `io.Discard`, particularly in non-hot code paths like these error handling bits. Then again I'm not sure being able to reuse the connection would prove to be a useful optimization for those same non-hot code paths

Reply:
Yeah, I think it wouldn't hurt but as you say it is not a big deal. I think we could leave it and maybe change it when touching this code again at some point.